parquet-pig/src/main/java/org/apache/parquet/pig/ParquetLoader.java

```diff
@@ -380,4 +397,133 @@ private Schema getSchemaFromRequiredFieldList(Schema schema, List<RequiredField>
+      switch(field.type) {
+        case DataType.BOOLEAN:
+        case DataType.INTEGER:
+        case DataType.LONG:
+        case DataType.FLOAT:
+        case DataType.DOUBLE:
+        case DataType.CHARARRAY:
+          fields.add(field.alias);
+          break;
+        default:
+          // Skip BYTEARRAY, TUPLE, MAP, BAG, DATETIME, BIGINTEGER, BIGDECIMAL
+          break;
+      }
+    }
+
+    return fields;
+  }
+
+  @Override
+  public List<Expression.OpType> getSupportedExpressionTypes() {
+    Expression.OpType supportedTypes [] = {
+        Expression.OpType.OP_EQ,
+        Expression.OpType.OP_GT,
+        Expression.OpType.OP_GE,
+        Expression.OpType.OP_LT,
+        Expression.OpType.OP_LE,
+        Expression.OpType.OP_AND,
+        Expression.OpType.OP_OR,
+    };
+
+    return Arrays.asList(supportedTypes);
+  }
+
+  @Override
+  public void setPushdownPredicate(Expression e) throws IOException {
+    LOG.info("Pig pushdown expression: " + e);
+
+    FilterPredicate pred = buildFilter(e);
+    LOG.info("Parquet filter predicate expression: " + pred);
+
+    storeInUDFContext(ParquetInputFormat.FILTER_PREDICATE, pred);
+  }
+
+  private FilterPredicate buildFilter(Expression e) {
+    if (e instanceof BinaryExpression) {
+      Expression lhs = ((BinaryExpression) e).getLhs();
+      Expression rhs = ((BinaryExpression) e).getRhs();
+
+      switch (e.getOpType()) {
+        case OP_EQ: return eq(getColumn((Column) lhs), getValue(rhs));
+        case OP_GT: return gt(getColumn((Column) lhs), getValue(rhs));
+        case OP_GE: return gtEq(getColumn((Column) lhs), getValue(rhs));
+        case OP_LT: return lt(getColumn((Column) lhs), getValue(rhs));
+        case OP_LE: return ltEq(getColumn((Column) lhs), getValue(rhs));
+        case OP_AND: return and(buildFilter(lhs), buildFilter(rhs));
+        case OP_OR: return or(buildFilter(lhs), buildFilter(rhs));
+      }
+    }
+
+    return null;
+  }
+
+  private ColumnWrapper getColumn(Column c) {
+    String column = c.getName();
+
+    try {
+      FieldSchema f = schema.getField(column);
+
+      switch (f.type) {
+        case DataType.BOOLEAN: return new ColumnWrapper<Binary>(binaryColumn(f.alias));
+        case DataType.INTEGER: return new ColumnWrapper<Integer>(intColumn(f.alias));
+        case DataType.LONG: return new ColumnWrapper<Long>(longColumn(f.alias));
+        case DataType.FLOAT: return new ColumnWrapper<Float>(floatColumn(f.alias));
+        case DataType.DOUBLE: return new ColumnWrapper<Double>(doubleColumn(f.alias));
+        case DataType.CHARARRAY: return new ColumnWrapper<Binary>(binaryColumn(f.alias));
+      }
+
+    } catch (FrontendException e) {
+      throw new RuntimeException("Error processing pushdown for column:" + c , e);
+    }
+
+    return null;
+  }
+
+  private static class ColumnWrapper<T extends Comparable<T>> extends Operators.Column<T> implements Operators.SupportsLtGt, Serializable {
```

Comment:
Yeah, the issue is getting the type system to recognize it because you can't use `T extends Column & SupportsLtGt` in some places. You can declare a method with a type parameter, but T is distinct from IntColumn because Java infers the concrete type of T from the invocation context so casting IntColumn to T requires a cast. You can, however, pass IntColumn to such a method.

The easiest way is to use the ColumnWrapper, but I have a branch that gets around it: https://github.com/rdblue/parquet-mr/commit/a39fdffcf171b0806c71cb9d0d78b05dac609d38#diff-19a7f147ae735f66dcb5f580ea724754R521